source/opt/instruction.cpp

```diff
@@ -530,7 +530,8 @@ void Instruction::UpdateLexicalScope(uint32_t scope) {
     i.dbg_scope_.SetLexicalScope(scope);
   }
   if (!IsLineInst() &&
-      context()->AreAnalysesValid(IRContext::kAnalysisDebugInfo)) {
+      context()->AreAnalysesValid(IRContext::kAnalysisDebugInfo) &&
```

Comment:
Since we invalidate analyses when we enable/disable them, could we modify AreAnalysesValid() to return
(valid_analyses_ & ~disabled_analyses_) == set ?

- running analysis doesn't check if it is enabled or not:
 * call DisableAnalysis(kAnalysisDefUse) -> kAnalysisDefUse is marked as invalid, and disabled.
 * call get_def_use_mgr() will run the analysis, and mark it as valid.

Reply:
I guess if you do that it would also be nice to debug assert() in places like get_def_use_mgr() that analysis is not disabled? (Or we can rely on std::unique to throw)

Reply:
> could we modify AreAnalysesValid() to return (valid_analyses_ & ~disabled_analyses_) == set ?

This does not work. Imagine X and Y are valid and not disabled. AreAnalysesValid(X) should be true but your code would return false. I think the existing code is fine. If you see a problem, please give an example.

> Running analysis doesn't check if it is enabled or not

Yes, but the current mechanism (checking if valid) has the same problems and is less intuitive.

> I guess if you do that it would also be nice to debug assert() in places like get_def_use_mgr() that analysis is not disabled?

Yes. I would add that if you and @s-perron approve this approach.